dsdl_compiler/README.md

```diff
@@ -0,0 +1,108 @@
+# UAVCAN DSDL compiler for the libcanard
+
+## Overview
+Libcanard_dsdlc is a tool to generate UAVCAN DSDL definitions to libcanard compatible message c-modules.
+
+Modules have: defines, enums, unions, structs, and encoding and decoding functions as defined in UAVCAN DSDL. Encoding and decoding functions use the decode and encode functions of libcanard for bit packing / unpacking.
+
+In c there is no namespace, so all the generated #define and function names are long having full folder path included. This is made to prevent the collision with each other and to the rest of the system.
+
+## Install & Integration
+To get libcanard from the git, make sure all the submodules are fetched too.
+
+`git submodules update --init --recursive`
+
+### Generating files
+`python3 libcanard_dsdlc --outdir <outdir> <dsdl-definition-uavcan-folder>`
+
+NOTE: If python2 is used, monotonic library is needed.
+
+`pip install monolith`
+
+### Using generated c-modules
+Include all or only selected message c-files to your compiler script (e.g. Makefile). Add include path to root of the <dsdl-generate-output-folder>.
+
+NOTE: compiled *.o files can't be compiled into to flat "build" directory as some files in DSDL have the same name.
+
+### Float16
+Generated structs in modules use float16 type when so specified in DSDL.
+In c-modules float16 is defined as uint16_t which can be overridden by defining FLOAT16 to something else e.g. __fp16.
+
+e.g. in Makefile
+
+`
+CFLAGS += -DFLOAT16=__fp16
+`
+
+## Using generated modules
+
+#### Encode NodeStatus-broadcast message
+```
+ #include "uavcan/protocol/NodeStatus.h"
+
+    /* Reserve memory and struct for messages */
+    uint8_t packed_uavcan_msg_buf[UAVCAN_PROTOCOL_NODESTATUS_MAX_SIZE];
+    /* MAX_SIZE comes from module header as pre-calculated */
+    uavcan_protocol_NodeStatus msg;
+
+    msg.uptime_sec = GetUptime();
+
+    msg.health = HEALTH_OK;
+    msg.mode = MODE_OPERATIONAL;
+
+    msg.sub_mode = sub_mode;
+    msg.vendor_specific_status_code = vendor_status_code;
+
+    /* Encode filled struct to packed_uavcan_msg_buf, ready to be send */
+    uint32_t len_of_packed_msg = uavcan_protocol_NodeStatusEncode(&msg, packed_uavcan_msg_buf, 0, 1);
+
+    canardBroadcast(&g_canard,
+                    UAVCAN_PROTOCOL_NODESTATUS_SIGNATURE,
+                    UAVCAN_PROTOCOL_NODESTATUS_ID,
+                    &g_bc_node_status_transfer_id,
+                    CANARD_TRANSFER_PRIORITY_MEDIUM,
+                    packed_uavcan_msg_buf,
+                    len_of_packed_msg);
+```
+
+All the *Encode calls need to have offset given as 0 and root item as 1 as parameters. This is needed, as the generated code may use offset and root item info in following *Encode calls.
```

Comment:
From and API design point of view, I am not a huge fan of this.

Maybe it would be better to rename the current function `uavcan_protocol_NodeStatusEncode_internal` and have a public API that does not expose offset and root, i.e.:

```cpp
uint32_t uavcan_protocol_NodeStatusEncode(uavcan_protocol_NodeStatus *source, void *msg_buf)
{
    return uavcan_protocol_NodeStatusEncode_internal(source, buf, 0, 1);
}
```

I think it makes the API more clear, and it will be harder to cleanup the API later, when consumers of this code are in the wild.

Reply:
I hate this API too and I like your idea. I'll check the flash size cost, but most likely this is the way to "fix" this. Thanks!